### ISSUE TITLE
All entity attributes gone

### ISSUE THREAD

**Issue by dchacke** [replikativ/datahike#111]
I have been using datahike in production where I write to disk. After restarting the server and reloading the db from disk, I got the following error while running the schema:

```clojure
Exception #error {
 :cause Cannot add #datahike/Datom [1 :db/ident :post/id 536871406 true] because of unique constraint: (#datahike/Datom [1 :db/ident :post/id 536870913 true])
 :data {:error :transact/unique, :attribute :db/ident, :datom #datahike/Datom [1 :db/ident :post/id 536871406 true]}
 :via
 [{:type clojure.lang.ExceptionInfo
   :message Cannot add #datahike/Datom [1 :db/ident :post/id 536871406 true] because of unique constraint: (#datahike/Datom [1 :db/ident :post/id 536870913 true])
   :data {:error :transact/unique, :attribute :db/ident, :datom #datahike/Datom [1 :db/ident :post/id 536871406 true]}
   :at [datahike.db$validate_datom invokeStatic db.cljc 955]}]
 :trace
 [[datahike.db$validate_datom invokeStatic db.cljc 955]
  [datahike.db$validate_datom invoke db.cljc 951]
  [datahike.db$with_datom invokeStatic db.cljc 1072]
  [datahike.db$with_datom invoke db.cljc 1071]
  [clojure.lang.AFn applyToHelper AFn.java 156]
  [clojure.lang.AFn applyTo AFn.java 144]
  [clojure.core$apply invokeStatic core.clj 667]
  [clojure.core$update_in$up__6838 invoke core.clj 6185]
  [clojure.core$update_in invokeStatic core.clj 6186]
  [clojure.core$update_in doInvoke core.clj 6172]
  [clojure.lang.RestFn invoke RestFn.java 467]
  [datahike.db$transact_report invokeStatic db.cljc 1116]
  [datahike.db$transact_report invoke db.cljc 1114]
  [datahike.db$transact_add invokeStatic db.cljc 1216]
  [datahike.db$transact_add invoke db.cljc 1198]
  [datahike.db$transact_tx_data invokeStatic db.cljc 1434]
  [datahike.db$transact_tx_data invoke db.cljc 1280]
  [datahike.core$with invokeStatic core.cljc 231]
  [datahike.core$with invoke core.cljc 224]
  [datahike.core$_transact_BANG_$fn__29220 invoke core.cljc 438]
  [clojure.lang.Atom swap Atom.java 37]
  [clojure.core$swap_BANG_ invokeStatic core.clj 2352]
  [clojure.core$swap_BANG_ invoke core.clj 2345]
  [datahike.core$_transact_BANG_ invokeStatic core.cljc 437]
  [datahike.core$_transact_BANG_ invoke core.cljc 434]
  [datahike.core$transact_BANG_ invokeStatic core.cljc 529]
  [datahike.core$transact_BANG_ invoke core.cljc 444]
  [datahike.core$transact invokeStatic core.cljc 636]
  [datahike.core$transact invoke core.cljc 629]
  [datahike.core$transact invokeStatic core.cljc 633]
  [datahike.core$transact invoke core.cljc 629]
  [datahike.connector$eval39584$fn__39586$fn__39588 invoke connector.cljc 31]
  [clojure.core$binding_conveyor_fn$fn__5739 invoke core.clj 2030]
  [clojure.lang.AFn call AFn.java 18]
  [java.util.concurrent.FutureTask run FutureTask.java 266]
  [java.util.concurrent.ThreadPoolExecutor runWorker ThreadPoolExecutor.java 1142]
  [java.util.concurrent.ThreadPoolExecutor$Worker run ThreadPoolExecutor.java 617]
  [java.lang.Thread run Thread.java 745]]}
```

It looks like it's trying to create two posts with the same db id *while loading the schema*. So that's the first weird thing; and my question about it is: am I not supposed to transact the schema once more against an existing database? Currently, I just transact the schema whether the db already exists or not.

The second weird thing (not sure if related or not) is that when I run a query like this one:

```clojure
(db/query '[:find [(pull ?e [*]) ...] :where [?e :notification/id]])
```

some of the query results look like this:

```clojure
#:db{:id 60}
```

Not all of them look like that - some of them do contain the `:notification/id` the query was looking for - but the odd thing is that the query would even return those fields that only contain the `:db/id`. After all, they don't match the query. Or at least they don't look like they do, though I have a hunch they *do* still contain all the other fields and that's they are included in the query results, but they aren't being displayed for some reason.

In other words, for a large part of the entities in my db, none of their attributes other than their `:db/id` are fetched; for those, all other entity attributes seem to be gone, or at least not fetched from the db.

**Comment by dchacke** [comment]
BTW on that note, shouldn't the schema be loaded *before* the rest of the db is loaded from disk?

**Comment by kordano** [comment]
thanks @dchacke for the insight. Usually you only need to write the schema once to the database and add new schemata afterwards. I'll have to take a look if somehow the read after re-connecting has some issues with some schemata.

**Comment by kordano** [comment]
Could you run me through the process how you read the db again from disk? I'm trying to understand how this could have happened.

**Comment by dchacke** [comment]
To read the db from disk, I do the following:

```clojure
(def uri "datahike:file:///tmp/diskussion-db")

(defn setup
  "Creates the db, connects to it, and
  adds the schema to the db."
  []
  (try
    (defonce db (d/create-database uri))
    (catch Exception e
      (println "Reusing db at" uri)))
  (defonce conn (d/connect uri))
  (transact schema))

(setup)
```

As you can see, this currently includes the mistaken step of transacting the schema every time.

**Comment by dchacke** [comment]
To read the db from disk, I do the following:

```clojure
(def uri "datahike:file:///path-to-db")

(defn setup
  "Creates the db, connects to it, and
  adds the schema to the db."
  []
  (try
    (defonce db (d/create-database uri))
    (catch Exception e
      (println "Reusing db at" uri)))
  (defonce conn (d/connect uri))
  (transact schema))

(setup)
```

As you can see, this currently includes the mistaken step of transacting the schema every time.

**Comment by kordano** [comment]
Yes, thanks. So you call this each time you restart your application with all data at that folder? Basically you shouldn't create the database again. A re-connection should be enough.

**Comment by dchacke** [comment]
> So you call this each time you restart your application with all data at that folder?

Yes.

> Basically you shouldn't create the database again. A re-connection should be enough.

Well, `d/create-database` throws an exception, which I use to infer that the database already exists. I wouldn't want to manually check for the existence of the database every time I run my application in a new environment. Is there a `(d/exists? uri)` or something along those lines that I could use instead of relying on the exception?

And: does `d/create-database` have an effect on the already existing db when it throws an exception because the db already exists?

**Comment by kordano** [comment]
There is a function called `database-exists?` in the `api`namespace for this kind of initialization.  `create-database` doesn't change anything if a store already exists in the location. Could you help me create a test case for your problem, so I can check it myself?
